dartagnan/src/main/java/com/dat3m/dartagnan/expression/INonDet.java

```diff
@@ -41,20 +44,26 @@ public IConst reduce() {
 	}
 
 	@Override
-	public Expr toZ3Int(Event e, Context ctx) {
+	public Formula toIntFormula(Event e, SolverContext ctx) {
 		String name = Integer.toString(hashCode());
-		return precision > 0 ? ctx.mkBVConst(name, precision) : ctx.mkIntConst(name);
+		FormulaManager fmgr = ctx.getFormulaManager();
+		return precision > 0 ? 
+				fmgr.getBitvectorFormulaManager().makeVariable(precision, name) : 
+				fmgr.getIntegerFormulaManager().makeVariable(name);
 	}
 
 	@Override
-	public Expr getLastValueExpr(Context ctx) {
+	public Formula getLastValueExpr(SolverContext ctx) {
 		String name = Integer.toString(hashCode());
-		return precision > 0 ? ctx.mkBVConst(name, precision) : ctx.mkIntConst(name);
+		FormulaManager fmgr = ctx.getFormulaManager();
+		return precision > 0 ? 
+				fmgr.getBitvectorFormulaManager().makeVariable(precision, name) : 
+				fmgr.getIntegerFormulaManager().makeVariable(name);
 	}
 
 	@Override
-	public BigInteger getIntValue(Event e, Model model, Context ctx) {
-		return new BigInteger(model.getConstInterp(toZ3Int(e, ctx)).toString());
+	public BigInteger getIntValue(Event e, Model model, SolverContext ctx) {
+		return new BigInteger(model.evaluate(toIntFormula(e, ctx)).toString());
```

Comment:
model.evaluate(...) may return null if the variable is not present in the model. It might be better to perform a check against that instead of letting the code run into a NullPointerException. It might not be necessary as the NullPointerException is thrown at the place of error (instead of some later time) so it is easy to debug.